include/libtorrent/aux_/rtc_stream.hpp

```diff
@@ -0,0 +1,343 @@
+	{ /* dummy */ }
+
+	template <class Mutable_Buffers>
+	std::size_t read_some(Mutable_Buffers const& buffers, error_code& ec)
+	{
+		TORRENT_ASSERT(!m_read_handler);
+		if (!is_open())
+		{
+			ec = boost::asio::error::not_connected;
+			return 0;
+		}
+
+		if (available() == 0)
+		{
+			ec = boost::asio::error::would_block;
+			return 0;
+		}
+#if TORRENT_USE_ASSERTS
+		size_t buf_size = 0;
+#endif
+
+		for (auto i = buffer_sequence_begin(buffers)
+			, end(buffer_sequence_end(buffers)); i != end; ++i)
+		{
+			add_read_buffer(i->data(), i->size());
+#if TORRENT_USE_ASSERTS
+			buf_size += i->size();
+#endif
+		}
+		std::size_t ret = read_some(true);
+		TORRENT_ASSERT(ret <= buf_size);
+		TORRENT_ASSERT(ret > 0);
+		return ret;
+	}
+
+	template <class Const_Buffers>
+	std::size_t write_some(Const_Buffers const& /* buffers */, error_code& /* ec */)
+	{
+		TORRENT_ASSERT(false && "not implemented!");
+		// TODO: implement blocking write (same as utp_stream)
+		return 0;
+	}
+
+#ifndef BOOST_NO_EXCEPTIONS
+	template <class Mutable_Buffers>
+	std::size_t read_some(Mutable_Buffers const& buffers)
+	{
+		error_code ec;
+		std::size_t ret = read_some(buffers, ec);
+		if (ec)
+			boost::throw_exception(boost::system::system_error(ec));
+		return ret;
+	}
+
+	template <class Const_Buffers>
+	std::size_t write_some(Const_Buffers const& buffers)
+	{
+		error_code ec;
+		std::size_t ret = write_some(buffers, ec);
+		if (ec)
+			boost::throw_exception(boost::system::system_error(ec));
+		return ret;
+	}
+#endif
+
+	template <class Const_Buffers, class Handler>
+	void async_write_some(Const_Buffers const& buffers, Handler const& handler)
+	{
+		if (!m_data_channel)
+		{
+			post(m_io_context, std::bind<void>(handler
+				, boost::asio::error::not_connected, std::size_t(0)));
+			return;
+		}
+
+		TORRENT_ASSERT(!m_write_handler);
+		if (m_write_handler)
+		{
+			post(m_io_context, std::bind<void>(handler
+				, boost::asio::error::operation_not_supported, std::size_t(0)));
+			return;
+		}
+
+		std::size_t bytes_added = 0;
+		for (auto i = buffer_sequence_begin(buffers)
+			, end(buffer_sequence_end(buffers)); i != end; ++i)
+		{
+			if (i->size() == 0) continue;
+			add_write_buffer(i->data(), i->size());
+			bytes_added += i->size();
+		}
+		if (bytes_added == 0)
+		{
+			// if we're writing 0 bytes, post handler immediately
+			// asio's SSL layer depends on this behavior
+			post(m_io_context, std::bind<void>(handler, error_code(), std::size_t(0)));
+			return;
+		}
+		m_write_handler = handler;
+		issue_write();
+	}
+
+private:
+	void on_message(error_code const& ec, std::vector<char> data);
+	void cancel_handlers(error_code const&);
+	bool ensure_open();
+
+	std::function<void(error_code const&, std::size_t)> m_read_handler;
+	std::function<void(error_code const&, std::size_t)> m_write_handler;
+
+	io_context& m_io_context;
+	std::shared_ptr<rtc::PeerConnection> m_peer_connection;
+	std::shared_ptr<rtc::DataChannel> m_data_channel;
```

Comment:
from the names, it sounds like `PeerConnection` represents the socket that's connected to the peer. but what is the `DataChannel` then? I would have expected that to be the "protocol" layered on top of the peer connection socket.
Please excuse my ignorance of RTC terminology if this is obvious.

Reply:
No, that's a good question. The libdatachannel library somewhat replicates the way the WebRTC API works in browsers. The `PeerConnection` represents the local endpoint of the connection, it handles negociation and abstracts the network stack. Then, different usages can be made of the `PeerConnection`, one of them being opening one or multiple `DataChannel` (abstracting SCTP channels under the hood), which expose a message socket API. Therefore, you need both to have the equivalent of an actual socket.

Reply:
In `rtc_stream`, the `PeerConnection` instance is mostly needed to retrieve the endpoints.